Fix UndoRedo::Operation objects cleanup

rxlecky opened the pull request:
Clean up `Reference` objects using `unref` rather than `memdelete` to prevent double free in `UndoRedo::create_action`

Fixes the second listed case in #50769

RandomShaper commented:
Looks good. Howerver, I think this is a good opportunity to improve this further.

I think that in every case of `Operation` deletion this is enough:
```
						if (!E->get().ref.is_valid()) {
							if (Object *obj = ObjectDB::get_instance(E->get().object)) { // Keeping this shortcut syntax is up to you :)
								memdelete(obj);
							}
						}
```

In other words (or, well, in words), the `Ref` in `Operation` will decrease the reference count automatically.

I'm 99% sure this is accurate. If it makes sense to you, please update your PR changing all cases and test it a bit. Otherwise, I can do a PR with that when this is merged as it is now.

rxlecky commented:
@RandomShaper that's a good point! I used your code and took it one step further, extracting the clean-up code into `UndoRedo::Operation` destructor so that all the duplicate clean-up code can be deleted. Let me know what you think.

RandomShaper reviewed (approved):
In Godot generally classes are either PODs or full fledged, and in this case `Operation` will not clearly be either of those. That said, I think that anyone maintaining this code in the future will be easily able to notice that `Operation` is treated as POD type throught the code –so it doesn't need initializers (`type` would be actually the only uninitialized member)– or any other "decoration" and that the destructor is exceptionally included for convenience. Considering such convenience is huge, I strongly believe that this is the right way to go in this case.

RandomShaper reviewed (changes requested):
I'm changing my mind in only in one thing: let's add an `UNINITIALIZED` item to the `Type` enumeration and have `type` initialized to that.
This is to avoid relying on `Vector`'s implementation details (we can't be sure it won't instantiate an `Operation` that will have `type` garbage-initialized and will be destroyed). No big deal, but just to be academically correct in that no uninitialized data is ever accessed.

RandomShaper commented:
You can just add this into the `switch`:
```
default: {
    CRASH_NOW();
}
```

rxlecky commented:
Added the default enumerator now. Only I named it `TYPE_INVALID` to make it consistent with other enumerators. Also since you mentioned the POD and full class types in Godot, I reverted `Operation` to use `struct` keyword to reinforce that it should be treated as POD type.

rxlecky commented:
Is crashing necessary? If there was a bug and the invalid operation type would make it into the list, invoking any undo/redo would crash the editor. Surely better user experience would be to print an error, saying there is an internal error that should be reported but the undo/redo would just not do anything in that case.

RandomShaper commented:
> Is crashing necessary?

Nope. It's just what I would have done, to signal a developer mistake. That said, what you've done is just as fine (or more).

RandomShaper commented:
Well, you still need to add the new case to the `switch`. I misread the code and thought you added the ' // do nothing` comment for `TYPE_INVALID`. Just testimonially handle it as you wish to get rid of the warning.

rxlecky commented:
Yes, sorry, I forgot handling all switch cases was required (not that it's a bad thing!).

RandomShaper reviewed (approved):
It's super good now! Thanks for your patience.

rxlecky commented:
Thank you for your feedback 🙂 It was certainly worth the extra effort and a good learning opportunity too.

akien-mga commented on core/undo_redo.cpp:
```diff
@@ -33,27 +33,21 @@
+UndoRedo::Operation::~Operation() {
+	if (type == Operation::TYPE_REFERENCE) {
+		if (!ref.is_valid()) {
+			if (Object *obj = ObjectDB::get_instance(object)) {
+				memdelete(obj);
+			}
```
We prefer this style:
```suggestion
			Object *obj = ObjectDB::get_instance(object);
			if (obj) {
				memdelete(obj);
			}
```
It's maybe a tad less elegant but creates no risk of misunderstanding.

rxlecky replied:
Fair point 🙂  Fixed now!

RandomShaper replied:
That's true today... But we'll have to discuss it at some point! 😃

rxlecky committed:
Implement UndoRedo::Operation destructor

This way we can delete all occurences of nearly identical cleanup code
scattered around the UndoRedo class that was error-prone to keep in sync.

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.4.1.

akien-mga commented:
Reverting for now with #55022 due to regressions (#55014 and #55015). But please feel free to open a new PR with this fix + the regressions fixed.
